src/cpp/session/modules/SessionThemes.cpp

```diff
@@ -362,6 +363,83 @@ Error getThemes(const json::JsonRpcRequest& request,
    return Success();
 }
 
+/**
+ * @brief RPC that lets the client add a theme for the current user.
+ *
+ * @param request       The request from the client to add a theme. The theme should already exist
+ *                      on the server and the only parameter on the request should be the location
+ *                      of the theme to add.
+ * @param pResponse     The response from the server. Will contain the name of the newly added theme.
+ *
+ * @return `Success` on success; an error otherwise.
+ */
+Error addTheme(const json::JsonRpcRequest& request,
+                     json::JsonRpcResponse* pResponse)
+{
+   std::string themeToAdd = request.params.at(0).get_str();
+
+   boost::regex rsThemeRegex(".*\\.rstheme$", boost::regex::icase);
+   boost::regex tmThemeRegex(".*\\.tmTheme$", boost::regex::icase);
+
+   Error error = Success();
+
+   // Find out whether to convert or add.
+   std::string funcName = ".rs.convertTheme";
+   bool isConversion = true;
+   if (boost::regex_match(themeToAdd, rsThemeRegex))
+   {
+      funcName = ".rs.addTheme";
+      isConversion = false;
+   }
+   else if (!boost::regex_match(themeToAdd, tmThemeRegex))
+   {
+      assert(false);
+      error = Error(json::errc::ParamInvalid, ERROR_LOCATION);
+      error.addProperty("queryParam", themeToAdd);
+      LOG_ERROR(error);
+   }
+
+   r::exec::RFunction rfunc(funcName);
+   rfunc.addParam("themePath", themeToAdd);
+   rfunc.addParam("apply", false);
+   rfunc.addParam("force", true);
+   rfunc.addParam("globally", false);
+
+   if (isConversion)
+   {
+      rfunc.addParam("add", true);
+      rfunc.addParam("outputLocation", R_NilValue);
+   }
+
+   SEXP result;
+   r::sexp::Protect protect;
+   error = rfunc.call(&result, &protect);
+   if (error)
+      LOG_ERROR(error);
```

Comment:
Should this error be returned from the RPC, too?

Reply:
It is, at line 422

Reply:
I missed that. Thanks!

Reply:
Np!